```
Error running mint automation
+ mkdir -p 13547-620b5a8/gopath/src/github.com/minio
+ git -C 13547-620b5a8/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 13547-620b5a8/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 13547-620b5a8/gopath/src/github.com/minio/minio fetch --quiet minio pull/13547/head:pr13547
+ git -C 13547-620b5a8/gopath/src/github.com/minio/minio checkout --quiet pr13547
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/13547-620b5a8/gopath
+ make -C 13547-620b5a8/gopath/src/github.com/minio/minio --quiet
go: downloading github.com/minio/cli v1.22.0
go: downloading cloud.google.com/go/storage v1.10.0
go: downloading github.com/Azure/azure-pipeline-go v0.2.2
go: downloading github.com/colinmarc/hdfs/v2 v2.2.0
go: downloading github.com/jcmturner/gokrb5/v8 v8.4.2
go: downloading google.golang.org/api v0.31.0
go: downloading github.com/minio/madmin-go v1.1.10
go: downloading github.com/minio/minio-go/v7 v7.0.15
go: downloading github.com/Azure/azure-storage-blob-go v0.10.0
go: downloading github.com/minio/pkg v1.1.5
go: downloading github.com/minio/highwayhash v1.0.2
go: downloading github.com/klauspost/compress v1.13.6
go: downloading github.com/dustin/go-humanize v1.0.0
go: downloading github.com/json-iterator/go v1.1.12
go: downloading github.com/fatih/color v1.13.0
go: downloading github.com/secure-io/sio-go v0.3.1
go: downloading github.com/minio/sha256-simd v1.0.0
go: downloading github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
go: downloading github.com/Shopify/sarama v1.27.2
go: downloading github.com/minio/sio v0.3.0
go: downloading github.com/golang-jwt/jwt v3.2.2+incompatible
go: downloading golang.org/x/crypto v0.0.0-20210921155107-089bfa567519
go: downloading github.com/minio/kes v0.14.0
go: downloading golang.org/x/sys v0.0.0-20211020174200-9d6173849985
go: downloading github.com/rs/dnscache v0.0.0-20210201191234-295bba877686
go: downloading github.com/cheggaaa/pb v1.0.29
go: downloading github.com/VividCortex/ewma v1.1.1
go: downloading github.com/minio/console v0.11.1-0.20211028015028-610c8a46533c
go: downloading github.com/bcicen/jstream v1.0.1
go: downloading github.com/bits-and-blooms/bloom/v3 v3.0.1
go: downloading github.com/djherbis/atime v1.0.0
go: downloading github.com/klauspost/reedsolomon v1.9.13
go: downloading github.com/google/uuid v1.3.0
go: downloading github.com/klauspost/readahead v1.3.1
go: downloading github.com/cespare/xxhash/v2 v2.1.2
go: downloading github.com/miekg/dns v1.1.35
go: downloading github.com/klauspost/pgzip v1.2.5
go: downloading github.com/go-openapi/loads v0.20.2
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/dchest/siphash v1.2.1
go: downloading github.com/cosnicolaou/pbzip2 v1.0.1
go: downloading github.com/minio/selfupdate v0.3.1
go: downloading github.com/pierrec/lz4 v2.6.0+incompatible
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/philhofer/fwd v1.1.1
go: downloading github.com/minio/zipindex v0.2.1
go: downloading github.com/prometheus/client_golang v1.11.0
go: downloading github.com/prometheus/client_model v0.2.0
go: downloading github.com/prometheus/procfs v0.7.3
go: downloading github.com/rs/cors v1.7.0
go: downloading github.com/shirou/gopsutil/v3 v3.21.9
go: downloading github.com/tinylib/msgp v1.1.6
go: downloading github.com/yargevad/filepathx v1.0.0
go: downloading github.com/valyala/bytebufferpool v1.0.0
go: downloading go.etcd.io/etcd/api/v3 v3.5.0
go: downloading go.etcd.io/etcd/client/v3 v3.5.0
go: downloading go.uber.org/atomic v1.9.0
go: downloading github.com/coredns/coredns v1.4.0
go: downloading github.com/beevik/ntp v0.3.0
go: downloading github.com/go-ldap/ldap/v3 v3.2.4
go: downloading go.uber.org/zap v1.19.1
go: downloading github.com/minio/simdjson-go v0.2.1
go: downloading github.com/montanaflynn/stats v0.6.6
go: downloading github.com/ncw/directio v1.0.5
go: downloading golang.org/x/time v0.0.0-20210220033141-f8bda1e9f3ba
go: downloading github.com/eclipse/paho.mqtt.golang v1.3.0
go: downloading github.com/elastic/go-elasticsearch/v7 v7.12.0
go: downloading github.com/cespare/xxhash v1.1.0
go: downloading github.com/go-sql-driver/mysql v1.5.0
go: downloading github.com/gomodule/redigo v2.0.0+incompatible
go: downloading github.com/lib/pq v1.9.0
go: downloading github.com/nats-io/nats.go v1.11.1-0.20210623165838-4b75fc59ae30
go: downloading github.com/nats-io/stan.go v0.8.3
go: downloading github.com/nsqio/go-nsq v1.0.8
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/streadway/amqp v1.0.0
go: downloading github.com/mattn/go-colorable v0.1.10
go: downloading github.com/mattn/go-isatty v0.0.14
go: downloading github.com/mattn/go-ieproxy v0.0.1
go: downloading github.com/bits-and-blooms/bitset v1.2.0
go: downloading golang.org/x/sync v0.0.0-20210220032951-036812b2e83c
go: downloading github.com/minio/csvparser v1.0.0
go: downloading github.com/minio/parquet-go v1.0.0
go: downloading github.com/go-ldap/ldap v3.0.2+incompatible
go: downloading github.com/alecthomas/participle v0.2.1
go: downloading go.etcd.io/etcd v0.5.0-alpha.5.0.20200910180754-dd1b699fc489
go: downloading github.com/golang/protobuf v1.5.2
go: downloading github.com/mattn/go-runewidth v0.0.13
go: downloading golang.org/x/net v0.0.0-20211020060615-d418f374d309
go: downloading github.com/xdg/stringprep v1.0.0
go: downloading github.com/klauspost/cpuid/v2 v2.0.9
go: downloading github.com/minio/argon2 v1.0.0
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/modern-go/reflect2 v1.0.2
go: downloading github.com/nats-io/nuid v1.0.1
go: downloading github.com/golang/snappy v0.0.3
go: downloading github.com/Azure/go-ntlmssp v0.0.0-20200615164410-66371956d46c
go: downloading github.com/go-asn1-ber/asn1-ber v1.5.1
go: downloading github.com/go-openapi/analysis v0.20.0
go: downloading github.com/go-openapi/spec v0.20.3
go: downloading github.com/go-openapi/swag v0.19.14
go: downloading github.com/rivo/uniseg v0.2.0
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading github.com/gorilla/websocket v1.4.2
go: downloading git.apache.org/thrift.git v0.13.0
go: downloading github.com/prometheus/common v0.31.1
go: downloading golang.org/x/text v0.3.7
go: downloading github.com/beorn7/perks v1.0.1
go: downloading github.com/jcmturner/dnsutils/v2 v2.0.0
go: downloading github.com/jcmturner/gofork v1.0.0
go: downloading github.com/hashicorp/go-uuid v1.0.2
go: downloading github.com/tidwall/gjson v1.10.2
go: downloading github.com/tidwall/sjson v1.1.6
go: downloading github.com/lestrrat-go/jwx v1.2.7
go: downloading github.com/rjeczalik/notify v0.9.2
go: downloading github.com/fatih/structs v1.1.0
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/shirou/gopsutil v0.0.0-20190901111213-e4ec7b275ada
go: downloading github.com/jcmturner/aescts/v2 v2.0.0
go: downloading github.com/jcmturner/rpc/v2 v2.0.3
go: downloading go.uber.org/multierr v1.7.0
go: downloading github.com/nats-io/nkeys v0.3.0
go: downloading github.com/mailru/easyjson v0.7.6
go: downloading github.com/rs/xid v1.3.0
go: downloading google.golang.org/protobuf v1.27.1
go: downloading gopkg.in/ini.v1 v1.63.2
go: downloading github.com/tidwall/pretty v1.2.0
go: downloading github.com/tidwall/match v1.1.1
go: downloading github.com/minio/md5-simd v1.1.2
go: downloading github.com/klauspost/cpuid v1.3.1
go: downloading cloud.google.com/go v0.65.0
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading gopkg.in/jcmturner/gokrb5.v7 v7.5.0
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading github.com/eapache/queue v1.1.0
go: downloading github.com/eapache/go-resiliency v1.2.0
go: downloading github.com/rcrowley/go-metrics v0.0.0-20200313005456-10cdbea86bc0
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.2-0.20181231171920-c182affec369
go: downloading github.com/josharian/intern v1.0.0
go: downloading github.com/jcmturner/goidentity/v6 v6.0.1
go: downloading github.com/go-openapi/strfmt v0.20.0
go: downloading github.com/go-openapi/jsonpointer v0.19.5
go: downloading github.com/tklauser/go-sysconf v0.3.9
go: downloading github.com/go-openapi/errors v0.19.9
go: downloading github.com/mitchellh/mapstructure v1.4.1
go: downloading go.mongodb.org/mongo-driver v1.4.6
go: downloading github.com/asaskevich/govalidator v0.0.0-20200907205600-7a23bdc65eef
go: downloading github.com/tklauser/numcpus v0.3.0
go: downloading github.com/lestrrat-go/option v1.0.0
go: downloading github.com/lestrrat-go/httpcc v1.0.0
go: downloading github.com/lestrrat-go/iter v1.0.1
go: downloading github.com/lestrrat-go/backoff/v2 v2.0.8
go: downloading github.com/lestrrat-go/blackmagic v1.0.0
go: downloading github.com/go-openapi/jsonreference v0.19.5
go: downloading github.com/PuerkitoBio/purell v1.1.1
go: downloading github.com/PuerkitoBio/urlesc v0.0.0-20170810143723-de5bf2ad4578
go: downloading google.golang.org/genproto v0.0.0-20210928142010-c7af6a1a74c9
go: downloading github.com/googleapis/gax-go/v2 v2.0.5
go: downloading go.opencensus.io v0.22.5
go: downloading google.golang.org/grpc v1.41.0
go: downloading github.com/googleapis/gax-go v2.0.2+incompatible
go: downloading github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e
go: downloading github.com/go-stack/stack v1.8.0
go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.0
go: downloading github.com/coreos/go-semver v0.3.0
go: downloading github.com/coreos/go-systemd/v22 v22.3.2
go: downloading github.com/coreos/go-systemd v0.0.0-20190719114852-fd7a80b32e1f
go: downloading github.com/go-openapi/runtime v0.19.24
go: downloading github.com/minio/mc v0.0.0-20211027024940-7866f97ef502
go: downloading github.com/jessevdk/go-flags v1.4.0
go: downloading github.com/minio/operator/logsearchapi v0.0.0-20211011212245-31460bbbc4b7
go: downloading github.com/unrolled/secure v1.0.9
go: downloading github.com/minio/operator v0.0.0-20211011212245-31460bbbc4b7
go: downloading github.com/minio/direct-csi v1.3.5-0.20210601185811-f7776f7961bf
go: downloading k8s.io/client-go v0.21.1
go: downloading golang.org/x/oauth2 v0.0.0-20210514164344-f6687ab2804c
go: downloading github.com/go-openapi/validate v0.20.2
go: downloading github.com/docker/go-units v0.4.0
go: downloading k8s.io/apimachinery v0.21.1
go: downloading k8s.io/kube-openapi v0.0.0-20210305001622-591a79e4bda7
go: downloading github.com/mb0/glob v0.0.0-20160210091149-1eb79d2de6c4
go: downloading github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
go: downloading github.com/briandowns/spinner v1.16.0
go: downloading github.com/minio/filepath v1.0.0
go: downloading github.com/pkg/xattr v0.4.3
go: downloading github.com/minio/colorjson v1.0.1
go: downloading github.com/posener/complete v1.2.3
go: downloading github.com/pkg/profile v1.6.0
go: downloading maze.io/x/duration v0.0.0-20160924141736-faac084b6075
go: downloading gopkg.in/h2non/filetype.v1 v1.0.5
go: downloading golang.org/x/term v0.0.0-20210927222741-03fcf44c2211
go: downloading github.com/hashicorp/go-multierror v1.1.1
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading github.com/georgysavva/scany v0.2.7
go: downloading k8s.io/klog/v2 v2.8.0
go: downloading k8s.io/api v0.21.1
go: downloading sigs.k8s.io/controller-runtime v0.8.0
go: downloading github.com/stretchr/testify v1.7.0
go: downloading k8s.io/klog v1.0.0
go: downloading github.com/go-logr/logr v0.4.0
go: downloading github.com/pmezard/go-difflib v1.0.1-0.20181226105442-5d4384ee4fb2
go: downloading github.com/stretchr/objx v0.2.0
go: downloading gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b
go: downloading sigs.k8s.io/yaml v1.2.0
go: downloading github.com/google/gofuzz v1.1.0
go: downloading sigs.k8s.io/structured-merge-diff/v4 v4.1.0
go: downloading github.com/google/go-cmp v0.5.5
go: downloading gopkg.in/inf.v0 v0.9.1
go: downloading sigs.k8s.io/structured-merge-diff v1.0.1-0.20191108220359-b1b620dd3f06
go: downloading github.com/emicklei/go-restful v2.9.5+incompatible
go: downloading k8s.io/utils v0.0.0-20201110183641-67b214c5f920
go: downloading github.com/googleapis/gnostic v0.5.1
go: downloading gopkg.in/jcmturner/dnsutils.v1 v1.0.1
go: downloading gopkg.in/jcmturner/aescts.v1 v1.0.1
go: downloading gopkg.in/jcmturner/rpc.v1 v1.1.0
cmd/api-errors.go:30:2: github.com/Azure/azure-storage-blob-go@v0.10.0: Get "https://proxy.golang.org/github.com/%21azure/azure-storage-blob-go/@v/v0.10.0.zip": dial tcp 142.250.189.241:443: connect: connection refused
cmd/object-api-utils.go:40:2: github.com/klauspost/readahead@v1.3.1: Get "https://proxy.golang.org/github.com/klauspost/readahead/@v/v1.3.1.zip": dial tcp 142.250.189.241:443: connect: connection refused
cmd/admin-handler-utils.go:25:2: github.com/minio/kes@v0.14.0: Get "https://proxy.golang.org/github.com/minio/kes/@v/v0.14.0.zip": dial tcp 142.250.189.241:443: connect: connection refused
cmd/s3-zip-handlers.go:36:2: github.com/minio/zipindex@v0.2.1: Get "https://proxy.golang.org/github.com/minio/zipindex/@v/v0.2.1.zip": dial tcp 142.250.189.241:443: connect: connection refused
make: *** [build] Error 1

```

--- a/.github/workflows/go-lint.yml
+++ b/.github/workflows/go-lint.yml
@@ -39,4 +39,5 @@ jobs:
           curl -L -o nancy https://github.com/sonatype-nexus-community/nancy/releases/download/${nancy_version}/nancy-${nancy_version}-linux-amd64 && chmod +x nancy
           go list -deps -json ./... | jq -s 'unique_by(.Module.Path)|.[]|select(has("Module"))|.Module' | ./nancy sleuth
           make
+          make test
           make test-race

--- a/cmd/admin-handlers-users-race_test.go
+++ b/cmd/admin-handlers-users-race_test.go
@@ -1,0 +1,126 @@
+// Copyright (c) 2015-2021 MinIO, Inc.
+//
+// This file is part of MinIO Object Storage stack
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// +build !race
+
+// Tests in this file are not run under the `-race` flag as they are too slow
+// and cause context deadline errors.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/minio/madmin-go"
+	minio "github.com/minio/minio-go/v7"
+)
+
+func runAllIAMConcurrencyTests(suite *TestSuiteIAM, c *check) {
+	suite.SetUpSuite(c)
+	suite.TestDeleteUserRace(c)
+	suite.TearDownSuite(c)
+}
+
+func TestIAMInternalIDPConcurrencyServerSuite(t *testing.T) {
+	testCases := []*TestSuiteIAM{
+		// Init and run test on FS backend with signature v4.
+		newTestSuiteIAM(TestSuiteCommon{serverType: "FS", signer: signerV4}),
+		// Init and run test on FS backend, with tls enabled.
+		newTestSuiteIAM(TestSuiteCommon{serverType: "FS", signer: signerV4, secure: true}),
+		// Init and run test on Erasure backend.
+		newTestSuiteIAM(TestSuiteCommon{serverType: "Erasure", signer: signerV4}),
+		// Init and run test on ErasureSet backend.
+		newTestSuiteIAM(TestSuiteCommon{serverType: "ErasureSet", signer: signerV4}),
+	}
+	for i, testCase := range testCases {
+		t.Run(fmt.Sprintf("Test: %d, ServerType: %s", i+1, testCase.serverType), func(t *testing.T) {
+			runAllIAMConcurrencyTests(testCase, &check{t, testCase.serverType})
+		})
+	}
+}
+
+func (s *TestSuiteIAM) TestDeleteUserRace(c *check) {
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	bucket := getRandomBucketName()
+	err := s.client.MakeBucket(ctx, bucket, minio.MakeBucketOptions{})
+	if err != nil {
+		c.Fatalf("bucket creat error: %v", err)
+	}
+
+	// Create a policy policy
+	policy := "mypolicy"
+	policyBytes := []byte(fmt.Sprintf(`{
+ "Version": "2012-10-17",
+ "Statement": [
+  {
+   "Effect": "Allow",
+   "Action": [
+    "s3:PutObject",
+    "s3:GetObject",
+    "s3:ListBucket"
+   ],
+   "Resource": [
+    "arn:aws:s3:::%s/*"
+   ]
+  }
+ ]
+}`, bucket))
+	err = s.adm.AddCannedPolicy(ctx, policy, policyBytes)
+	if err != nil {
+		c.Fatalf("policy add error: %v", err)
+	}
+
+	userCount := 50
+	accessKeys := make([]string, userCount)
+	secretKeys := make([]string, userCount)
+	for i := 0; i < userCount; i++ {
+		accessKey, secretKey := mustGenerateCredentials(c)
+		err = s.adm.SetUser(ctx, accessKey, secretKey, madmin.AccountEnabled)
+		if err != nil {
+			c.Fatalf("Unable to set user: %v", err)
+		}
+
+		err = s.adm.SetPolicy(ctx, policy, accessKey, false)
+		if err != nil {
+			c.Fatalf("Unable to set policy: %v", err)
+		}
+
+		accessKeys[i] = accessKey
+		secretKeys[i] = secretKey
+	}
+
+	wg := sync.WaitGroup{}
+	for i := 0; i < userCount; i++ {
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			uClient := s.getUserClient(c, accessKeys[i], secretKeys[i], "")
+			err := s.adm.RemoveUser(ctx, accessKeys[i])
+			if err != nil {
+				c.Fatalf("unable to remove user: %v", err)
+			}
+			c.mustNotListObjects(ctx, uClient, bucket)
+		}(i)
+	}
+	wg.Wait()
+}

--- a/cmd/admin-handlers-users_test.go
+++ b/cmd/admin-handlers-users_test.go
@@ -32,7 +32,7 @@ import (
 )
 
 const (
-	testDefaultTimeout = 10 * time.Second
+	testDefaultTimeout = 30 * time.Second
 )
 
 // API suite container for IAM

--- a/cmd/gateway-main.go
+++ b/cmd/gateway-main.go
@@ -306,7 +306,7 @@ func StartGateway(ctx *cli.Context, gw Gateway) {
 		logger.FatalIf(globalNotificationSys.Init(GlobalContext, buckets, newObject), "Unable to initialize notification system")
 	}
 
-	go globalIAMSys.Init(GlobalContext, newObject, globalEtcdClient)
+	go globalIAMSys.Init(GlobalContext, newObject, globalEtcdClient, globalRefreshIAMInterval)
 
 	if globalCacheConfig.Enabled {
 		// initialize the new disk cache objects.

--- a/cmd/iam.go
+++ b/cmd/iam.go
@@ -206,6 +206,8 @@ func newMappedPolicy(policy string) MappedPolicy {
 type IAMSys struct {
 	sync.Mutex
 
+	iamRefreshInterval time.Duration
+
 	usersSysType UsersSysType
 
 	// map of policy names to policy definitions
@@ -473,9 +475,9 @@ func (sys *IAMSys) Load(ctx context.Context, store IAMStorageAPI) error {
 	iamGroupPolicyMap := make(map[string]MappedPolicy)
 	iamPolicyDocsMap := make(map[string]iampolicy.Policy)
 
-	store.rlock()
+	store.lock()
+	defer store.unlock()
 	isMinIOUsersSys := sys.usersSysType == MinIOUsersSysType
-	store.runlock()
 
 	if err := store.loadPolicyDocs(ctx, iamPolicyDocsMap); err != nil {
 		return err
@@ -517,9 +519,6 @@ func (sys *IAMSys) Load(ctx context.Context, store IAMStorageAPI) error {
 	if err := store.loadMappedPolicies(ctx, stsUser, false, iamUserPolicyMap); err != nil {
 		return err
 	}
-
-	store.lock()
-	defer store.unlock()
 
 	for k, v := range iamPolicyDocsMap {
 		sys.iamPolicyDocsMap[k] = v
@@ -565,7 +564,9 @@ func (sys *IAMSys) Load(ctx context.Context, store IAMStorageAPI) error {
 }
 
 // Init - initializes config system by reading entries from config/iam
-func (sys *IAMSys) Init(ctx context.Context, objAPI ObjectLayer, etcdClient *etcd.Client) {
+func (sys *IAMSys) Init(ctx context.Context, objAPI ObjectLayer, etcdClient *etcd.Client, iamRefreshInterval time.Duration) {
+	sys.iamRefreshInterval = iamRefreshInterval
+
 	// Initialize IAM store
 	sys.InitStore(objAPI, etcdClient)
 
@@ -649,14 +650,14 @@ func (sys *IAMSys) Init(ctx context.Context, objAPI ObjectLayer, etcdClient *etc
 	case globalOpenIDConfig.ProviderEnabled():
 		go func() {
 			for {
-				time.Sleep(globalRefreshIAMInterval)
+				time.Sleep(sys.iamRefreshInterval)
 				sys.purgeExpiredCredentialsForExternalSSO(ctx)
 			}
 		}()
 	case globalLDAPConfig.EnabledWithLookupBind():
 		go func() {
 			for {
-				time.Sleep(globalRefreshIAMInterval)
+				time.Sleep(sys.iamRefreshInterval)
 				sys.purgeExpiredCredentialsForLDAP(ctx)
 				sys.updateGroupMembershipsForLDAP(ctx)
 			}
@@ -686,7 +687,7 @@ func (sys *IAMSys) watch(ctx context.Context) {
 	} else {
 		// Fall back to loading all items
 		for {
-			time.Sleep(globalRefreshIAMInterval)
+			time.Sleep(sys.iamRefreshInterval)
 			if err := sys.Load(ctx, sys.store); err != nil {
 				logger.LogIf(ctx, err)
 			}

--- a/cmd/server-main.go
+++ b/cmd/server-main.go
@@ -570,7 +570,7 @@ func serverMain(ctx *cli.Context) {
 	}
 
 	// Initialize users credentials and policies in background right after config has initialized.
-	go globalIAMSys.Init(GlobalContext, newObject, globalEtcdClient)
+	go globalIAMSys.Init(GlobalContext, newObject, globalEtcdClient, globalRefreshIAMInterval)
 
 	initDataScanner(GlobalContext, newObject)
 

--- a/cmd/test-utils_test.go
+++ b/cmd/test-utils_test.go
@@ -350,7 +350,7 @@ func UnstartedTestServer(t TestErrHandler, instanceType string) TestServer {
 
 	initAllSubsystems(ctx, objLayer)
 
-	globalIAMSys.Init(ctx, objLayer, globalEtcdClient)
+	globalIAMSys.Init(ctx, objLayer, globalEtcdClient, 2*time.Second)
 
 	return testServer
 }

